drivers/java/src/main/java/com/rethinkdb/ast/ReqlAst.java

```diff
@@ -96,6 +170,47 @@ protected Object build() {
     }
 
     /**
+     * Runs this query via connection {@code conn} provided by connection pool,
+     * with options {@code runOpts} and returns an atom result
+     * or a sequence result as a cursor. The atom result representing a JSON object is converted
+     * to an object of type {@code Class<P>} specified with {@code pojoClass}. The cursor
+     * is a {@code com.rethinkdb.net.Cursor} which may be iterated to get a sequence of atom results
+     * of type {@code Class<P>}
+     * @param pool The connection pool to provide connection to run this query
+     * @param pojoClass The class of POJO to convert to
+     * @param <T> The type of result
+     * @param <P> The type of POJO to convert to
+     * @return The result of this query (either a {@code P or a Cursor<P>}
+     */
+    public <T, P> T run(ConnectionPool pool, Class<P> pojoClass) {
+    	  Connection conn = pool.leaseConnection();
+    	  T res = this.run(conn, pojoClass);
+    	  pool.returnConnection(conn);
+    	  return res;
```

Comment:
`ReqlDriverError` is an unchecked exception extending `RuntimeException`, so user code can not really anticipate it or prepare to handle it. The docs and/or examples do not show any required handling in case of this exception (e.g. closing `Connection`).

Normally the connection pool wrapper code handles the lease-use-return pattern, except in cases of exception, which normally break the execution of current thread.

Java hint: unchecked exceptions (e.g. the ones extending `RuntimeException`) are the ones that you can not expect the calling code to recover from. If there is an anticipated or expected way to recover from the exception it should be a checked exception, where calling code is forced to deal with (at compile time).

`ConnectionPool` implementations pass this unchecked exception to the user code, the same as non-connection-pool methods.

Note: in case of exception the `Connection` is not returned back to connection pool and is subject to garbage collection.